### ISSUE TITLE
Backend_health reports a misleading error when poll(POLLIN) times out for a probe

### ISSUE THREAD

**Issue by slimhazard** [varnishcache/varnish-cache#2866]
When a backend probe polls the fd for read after sending the request in ``vbp_poke()`` (currently line 353 in ``cache_backend_probe.c``), the probe fails if ``poll(2)`` returns 0.

The error message ``Poll error`` is created, which is appended to ``Backend_health`` log entries together with ``errno`` and ``strerror(errno)``. But those always come out as 0 and "Success", because return value 0 indicates a timeout, and errno doesn't get set:
```
$ varnishlog -g raw -i Backend_health
0 Backend_health - varnish-8445d4f7f-n74dl Still sick 4---Xr-- 0 3 8 0.000000 0.000000 Poll error 0 (Success)
0 Backend_health - varnish-8445d4f7f-plcqt Still sick 4---Xr-- 0 3 8 0.000000 0.000000 Poll error 0 (Success)
0 Backend_health - varnish-8445d4f7f-n74dl Still sick 4---Xr-- 0 3 8 0.000000 0.000000 Poll error 0 (Success)
0 Backend_health - varnish-8445d4f7f-plcqt Still sick 4---Xr-- 0 3 8 0.000000 0.000000 Poll error 0 (Success)
```
According to both [Open Group](http://pubs.opengroup.org/onlinepubs/009695399/functions/poll.html) and poll(2) on my system, in "RETURN VALUE":

> A value of 0 indicates that the call timed out and no file descriptors have been selected.

So that's what the error message should say.

The code just below that in ``cache_backend_probe.c`` reports something about a timeout under another condition; that message may be getting it wrong as well.

Looks like this may have been introduced in 09a75daaf302eba6c4f81ff30f394f013d93f8c5.

**Comment by stefandjordjevichumanity** [comment]
Hi, pardon me if I by mistake hijack issue but seems that our current setup is affected by described problem which leads us to "No backed" and 503 in our performance testing environment with one backed server.

This is 100% reproducible with endurance test that we are running towards our api.
First 503 in "BerespStatus" confused me, but after some tcpdump-ing and error event analyses seems that  issue is that our backed is recognizes as sick, and happens pretty randomly in time during execution of this test. There is no threshold or limit hitting in setup (limits, open file/socket etc.) 

This is backend_fetch fail: 

`3080210 VCL_call       b BACKEND_FETCH
   3080210 VCL_return     b fetch
   3080210 FetchError     b Director acl1 returned no backend
   3080210 FetchError     b No backend
   3080210 Timestamp      b Beresp: 1546077500.373918 0.000009 0.000009
   3080210 Timestamp      b Error: 1546077500.373920 0.000011 0.000002
   3080210 BerespProtocol b HTTP/1.1
   3080210 BerespStatus   b 503
   3080210 BerespReason   b Service Unavailable
   3080210 BerespReason   b Backend fetch failed
`

And this is Backend_health:

`0 Backend_health - boot.app01 Still healthy 4---Xr-- 2 2 3 0.000000 0.001782 Poll error 0 (Success)
         0 Backend_health - boot.app01 Went sick 4---Xr-- 1 2 3 0.000000 0.001782 Poll error 0 (Success)
         0 Backend_health - boot.app01 Still sick 4---X-RH 1 2 3 0.001403 0.001687 HTTP/1.1 200 OK
         0 Backend_health - boot.app01 Back healthy 4---X-RH 2 2 3 0.003264 0.002081 HTTP/1.1 200 OK
         0 Backend_health - boot.app01 Still healthy 4---X-RH 3 2 3 0.003094 0.002334 HTTP/1.1 200 OK
`

So I did a testing on different versions of varnish. And results are:

varnish-4.0.1: test Fail-ing/issue with sick backend
varnish-4.1.10: test Pass-ing/no issue with sick backend 
varnish-5.1.3: test Pass-ing/no issue with sick backend
varnish-6.0.2: test Fail-ing/issue with sick backend

When it comes to behaviour with 503, not sure if this is a desired behaviour.
Thank you.

**Comment by slimhazard** [comment]
@stefandjordjevichumanity this issue is about the misleading error message -- likely a better error message would have helped you.

But it's not about the fact that the backends are found to be unhealthy, AFAICT that's working as designed. Since you're getting the "Poll error 0 (Success)" message, what's actually happening is that your backend ``app01`` is not responding to the health probe within the timeout set for the probe. You can also see that from the small "r" in ``4---Xr--``, which indicates an error receiving the response (see vsl(7) for details about reading the log entry).

I found that increasing the probe timeout (and the interval along with it) was enough to keep the backends healthy. Or you may need to fix the part of the app that responds to the probe. At any rate, those aren't Varnish bugs (except for the poor error message).

**Comment by slimhazard** [comment]
tbh I was surprised about the timeout that led to all of this, but didn't investigate it further after increasing the timeout made the problem go away.

@stefandjordjevichumanity in which Varnish version are you seeing this? In my case it was 6.1.1.

You may be right and there's *also* a problem with probes timing out when they shouldn't in certain Varnish versions. If that's so then we have new bug.

**Comment by stefandjordjevichumanity** [comment]
@slimhazard 
Thank you for quick response. 
Results I got, issue I see per varnish versions. 

varnish-4.0.1: FAIL-ing/issue with sick backend
varnish-4.1.10: PASS-ing/no issue with sick backend
varnish-5.1.3: PASS-ing/no issue with sick backend
varnish-6.0.2: FAIL-ing/issue with sick backend

What I just did, I executed test with: varnish-4.0.1 but I just increase timeout for 1s
```
.timeout = 3s;
```
Test is PASS-ing now but I caught only this:
```
0 Backend_health - app01 Still healthy 4--X-RH 3 2 3 0.001730 0.001587 HTTP/1.1 200 OK
0 Backend_health - app01 Still healthy 4--X--- 2 2 3 0.000000 0.001587 
0 Backend_health - app01 Still healthy 4--X-RH 2 2 3 0.388881 0.098411 HTTP/1.1 200 OK
```
Seems like one probe was out of timeout, or not sent at all, not sure if this is relevant catch.

I guess I try to bump timeout to a 1 minute if I will be able to catch that there was some "Response time" above 2 seconds (initial setup)

**Comment by stromnet** [comment]
Just a FYI, in case anyone else has similar issues:
This patch does not seem to be present in varnish-plus aka enterprise 6.0.6r7. Same behaviour observed, and cannot see reference to ticket in Changelog.

Probes where for some reason failing with `Error Recv` (from `varnishadm backend.list -p`), but no further details.
With the `varnishlog -g raw` trick (which itself took a while to find, I did not find any  apparent documentation on how to debug probe failures) I noticed the `Poll error 0 (Success)` message, which lead me here. 
Increasing the timeout immediately fixed the probe failures.
